### PR TITLE
Align the behaviour of `--path` in `tuist fetch` with the rest of the CLI

### PR DESCRIPTION
**Pull request by danibachar**
Resolves https://github.com/tuist/tuist/issues/5608

### Short description 📝

Enabling a smart search of the Dependencies.swift file for fetching using the `-p` flag from everywhere!

### How to test the changes locally 🧐

1. Run the TuistKitTests suite.
2. To use a Tuist build with this changes and test against the steps described in the issue:
- checkout this branch
- follow the steps in the issue above and see that they are resolved!
- Note for some reason building tuist from source does not recognize the ProjectDescriptionHelpers framework. As such comment out its import in the Dependencies.swift file (line 3) and either comment out line number 7 or bring the code from the ProjectDescriptionHelpers folder into the file
- After this modification biulding and running from source should work `swift build ProjectDescription && swift run tuist fetch -p fixtures/app_with_spm_dependencies/Features/FeatureOne`
- To generate the specific project run `swift run tuist generate -p fixtures/app_with_spm_dependencies/Features/FeatureOne`

### Contributor checklist ✅

- [x] The code has been linted using run `make workspace/lint-fix`
- [x] The change is tested via unit testing or acceptance testing, or both
- [x] The title of the PR is formulated in a way that is usable as a changelog entry
- [ ] In case the PR introduces changes that affect users, the documentation has been updated

### Reviewer checklist ✅

- [ ] The code architecture and patterns are consistent with the rest of the codebase
- [ ] Reviewer has checked that, if needed, the PR includes the label `changelog:added`, `changelog:fixed`, or `changelog:changed`, and the title is usable as a changelog entry

### CODE REVIEW

File: Sources/TuistKit/Services/FetchService.swift [hunk 1]
```diff
@@ -14,34 +14,65 @@ final class FetchService {
...
+        }
+
+        // Else, we try to traverse up the directories to find it
+        if let traversedPath = fileHandler.locateDirectoryTraversingParents(from: path, path: Constants.tuistDirectoryName) {
+            if fileHandler.exists(
+                traversedPath.appending(components: Constants.tuistDirectoryName, Manifest.dependencies.fileName(path))
+            ) {
+                return traversedPath
+            }
+        }
```
Comment (danyf90) [inline Sources/TuistKit/Services/FetchService.swift:70]: What is this for? I don't think we support intermediate Tuist folders, so it should be enough to always return the root

**Review by danyf90 — CHANGES_REQUESTED** [review]
Note that we are dropping support for Dependencies.swift in Tuist 4, would this stil be relevant? I would probably just do this PR on top of the PR removing it

File: Sources/TuistKit/Services/FetchService.swift (hunk 1 above)
Comment (danibachar) [reply Sources/TuistKit/Services/FetchService.swift:70]: I saw we support it with the [Config.swift ](https://github.com/tuist/tuist/blob/main/Sources/TuistLoader/Loaders/ConfigLoader.swift#L66) - I can remove if needed

**Comment by danibachar** [comment]
> Note that we are dropping support for Dependencies.swift in Tuist 4, would this still be relevant? I would probably just do this PR on top of the PR removing it

I'd love to see this issue fixed and released in the next version (I'm guessing there will be at least one more before Tuist 4?), I'm not 100% sure about the Tuist 4 release. This fix is crucial for adopting the Tuist cache (we have quite a complex modular setup).

File: Sources/TuistKit/Services/FetchService.swift (hunk 1 above)
Comment (pepicrft) [reply Sources/TuistKit/Services/FetchService.swift:70]: I agree with @danyf90. Let's remove that piece of code.

File: Tests/TuistKitTests/Services/FetchServiceTests.swift [hunk 2]
```diff
@@ -192,4 +192,118 @@ final class FetchServiceTests: TuistUnitTestCase {
 
         XCTAssertFalse(dependenciesController.invokedUpdate)
     }
+
+    func test_fetch_path_is_found_in_root() async throws {
+        // Given
+        let stubbedPath = try temporaryPath()
```
Comment (pepicrft) [inline Tests/TuistKitTests/Services/FetchServiceTests.swift:198]: This is not a `stubbedPath`, it's an actual directory. Let's name it `temporaryDirectory`.

File: Tests/TuistKitTests/Services/FetchServiceTests.swift [hunk 3]
```diff
@@ -192,4 +192,118 @@ final class FetchServiceTests: TuistUnitTestCase {
 
         XCTAssertFalse(dependenciesController.invokedUpdate)
     }
+
+    func test_fetch_path_is_found_in_root() async throws {
```
Comment (pepicrft) [inline Tests/TuistKitTests/Services/FetchServiceTests.swift:196]: `test_fetch_path_is_found_in_root` ~> `test_fetch_when_from_a_tuist_project_directory`

File: Tests/TuistKitTests/Services/FetchServiceTests.swift [hunk 4]
```diff
@@ -192,4 +192,118 @@ final class FetchServiceTests: TuistUnitTestCase {
...
+        }
+
+        // Dependencies.swift in root
+        try fileHandler.touch(expectedFoundDependenciesLocation)
+
+        // When
+        try await subject.run(
+            path: stubbedPath.pathString,
+            update: false
+        )
```
Comment (pepicrft) [inline Tests/TuistKitTests/Services/FetchServiceTests.swift:219]: What are we asserting here? I'd recommend being more explicit about what we expect, either through the test code or documentation comments. For example, if what we are expecting here is that the function succeeds, let's assert that it doesn't throw an error or add a comment saying that the function invocation should succeed.

Comment (pepicrft) [reply Tests/TuistKitTests/Services/FetchServiceTests.swift:219]: This comment applies to all the test functions in this class.

**Review by pepicrft — CHANGES_REQUESTED** [review]
Thanks a lot for tackling this @danibachar. I added some minor comments throughout the changes. Let me know when they are addressed and I'll give it a second look.

**Comment by pepicrft** [comment]
@all-contributors add @danibachar for code

File: Sources/TuistKit/Services/FetchService.swift (hunk 1 above)
Comment (danibachar) [reply Sources/TuistKit/Services/FetchService.swift:70]: Fixed

File: Tests/TuistKitTests/Services/FetchServiceTests.swift (hunk 4 above)
Comment (danibachar) [reply Tests/TuistKitTests/Services/FetchServiceTests.swift:219]: Thanks for this comment, and good callout, I forgot to add expectation.
The assertion takes place in the `loadDependenciesStub` closure.

I have added comments, and the expectation - please let me know if that is ok

**Comment by danibachar** [comment]
@pepicrft and @danyf90 - thank you both for your comments, I have fixed them (:

**Review by danyf90 — APPROVED** [review]

**Comment by danibachar** [comment]
I'm not sure why does the `Tuist / Cache warm with latest Tuist on Silicon (pull_request)` stage fails, It seems like a CI issue?
I updated my branch with main...

**Pull request merged**